File: Data Structure/328. Odd Even Linked List/328-solution.py
# https://leetcode.com/problems/odd-even-linked-list/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        ptr_odd = head
        odd_head = ptr_odd
        ptr_even = head.next
        even_head = ptr_even
        node = head.next.next
        odd_flag = True
        while not node:
            if odd_flag:
                odd_flag = False
                ptr_odd.next = node
                ptr_odd = ptr_odd.next
            else:
                odd_flag = True
                ptr_even.next = node
                ptr_even= ptr_even.next
            node = node.next
        ptr_odd.next = even_head
        return head

File: Data Structure/Linked List/328. Odd Even Linked List/328-solution.cpp
// https://leetcode.com/problems/odd-even-linked-list/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
  struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
  };

class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head == NULL || head->next == NULL)
            return head;
        auto odd_p = head;
        auto even_p = head->next;
        auto node = head->next->next;
        bool odd_f = 1;
        auto tmp = head->next;
        while(node != nullptr){
            if(odd_f){
                odd_p->next = node;
                odd_f = 0;
                odd_p = odd_p->next;
            }
            else{
                even_p->next = node;
                odd_f = 1;
                even_p = even_p->next;
            }
            node = node->next;
        }
        even_p->next = nullptr;
        odd_p->next = tmp;
        return head; 
    }
};